File: sdk/python/flagit/cached.py
"""The cached client: stream once, read locally, batch usage events."""

from __future__ import annotations

import threading
from typing import Any, Callable, Dict, Optional

from .client import Client, EvalContext, Evaluation, EvaluationDetail, FlagsState, Stream


class CachedClient:
    """Streams flag values into an in-memory store once, then serves synchronous
    local reads with ZERO network per read, reporting usage as batched summary
    events. Mirrors the LaunchDarkly-style pattern.

    Use this for a long-lived app; use ``Client`` for one-off calls. All methods
    are safe for concurrent use.
    """

    def __init__(
        self,
        base_url: str,
        sdk_key: str,
        context: EvalContext,
        flush_interval: float = 5.0,
        send_events: bool = True,
        on_error: Optional[Callable[[BaseException], None]] = None,
    ) -> None:
        self._base = Client(base_url, sdk_key)
        self._track = send_events
        self._on_error = on_error or (lambda _e: None)

        self._lock = threading.RLock()
        self._store: Dict[str, Evaluation] = {}
        self._ready = threading.Event()
        self._context = context
        self._stream: Optional[Stream] = None
        self._listeners: Dict[int, Callable[[], None]] = {}
        self._next_id = 0
        self._counts: Dict[str, Dict[int, int]] = {}
        self._closed = threading.Event()

        self._start_stream()

        self._timer: Optional[threading.Thread] = None
        if flush_interval > 0 and self._track:
            self._flush_interval = flush_interval
            self._timer = threading.Thread(target=self._flush_loop, daemon=True)
            self._timer.start()

    # -- streaming -----------------------------------------------------------

    def _start_stream(self) -> None:
        self._stream = self._base.stream(self._context, self._on_update, self._on_error)

    def _on_update(self, flags: Dict[str, Evaluation]) -> None:
        with self._lock:
            self._store = flags
            listeners = list(self._listeners.values())
        self._ready.set()
        for listener in listeners:
            listener()

    def _flush_loop(self) -> None:
        while not self._closed.wait(self._flush_interval):
            self.flush()

    # -- lifecycle -----------------------------------------------------------

    def wait_for_initialization(self, timeout: Optional[float] = None) -> bool:
        """Block until the initial snapshot arrives. Returns True if it did."""
        return self._ready.wait(timeout)

    def is_initialized(self) -> bool:
        """Whether the initial snapshot has loaded (matches LaunchDarkly's
        ``is_initialized``)."""
        return self._ready.is_set()

    def on_change(self, listener: Callable[[], None]) -> Callable[[], None]:
        """Subscribe to store updates; returns an unsubscribe function."""
        with self._lock:
            listener_id = self._next_id
            self._next_id += 1
            self._listeners[listener_id] = listener

        def unsubscribe() -> None:
            with self._lock:
                self._listeners.pop(listener_id, None)

        return unsubscribe

    # -- reads ---------------------------------------------------------------

    def _read(self, flag_key: str) -> Optional[Evaluation]:
        with self._lock:
            ev = self._store.get(flag_key)
            if ev is not None and self._track:
                by_var = self._counts.setdefault(flag_key, {})
                by_var[ev.variation] = by_var.get(ev.variation, 0) + 1
        return ev

    def bool_variation(self, flag_key: str, fallback: bool) -> bool:
        """Synchronous local read — no network. ``fallback`` before init / on mismatch."""
        ev = self._read(flag_key)
        return ev.value if ev is not None and isinstance(ev.value, bool) else fallback

    def string_variation(self, flag_key: str, fallback: str) -> str:
        ev = self._read(flag_key)
        return ev.value if ev is not None and isinstance(ev.value, str) else fallback

    def number_variation(self, flag_key: str, fallback: float) -> float:
        ev = self._read(flag_key)
        if ev is not None and isinstance(ev.value, (int, float)) and not isinstance(ev.value, bool):
            return ev.value
        return fallback

    def variation(self, flag_key: str, default: Any) -> Any:
        """The flag's value from the local store, or ``default``. The canonical
        read (matching LaunchDarkly's ``variation``); the typed helpers above are
        thin conveniences over it."""
        ev = self._read(flag_key)
        return ev.value if ev is not None else default

    def variation_detail(self, flag_key: str, default: Any) -> EvaluationDetail:
        """Like ``variation`` but also returns the variation index and reason
        (matching LaunchDarkly's ``variation_detail``)."""
        ev = self._read(flag_key)
        if ev is None:
            return EvaluationDetail(value=default, variation_index=None, reason={"kind": "FLAG_NOT_FOUND"})
        return EvaluationDetail(value=ev.value, variation_index=ev.variation, reason=ev.reason)

    def all_flags_state(self) -> FlagsState:
        """A snapshot of all cached evaluations. Returns a ``FlagsState``; call
        ``.to_values_dict()`` for a plain value map."""
        with self._lock:
            return FlagsState(dict(self._store))

    # -- usage events --------------------------------------------------------

    def flush(self) -> None:
        """Send buffered usage events now."""
        with self._lock:
            if not self._counts:
                return
            summary = {
                "flags": {
                    flag_key: {
                        "counters": [
                            {"variation": variation, "count": count}
                            for variation, count in by_var.items()
                        ]
                    }
                    for flag_key, by_var in self._counts.items()
                }
            }
            self._counts = {}
        try:
            self._base.send_events(summary)
        except Exception as err:  # noqa: BLE001 - surfaced via on_error
            self._on_error(err)

    def identify(self, context: EvalContext, timeout: Optional[float] = None) -> bool:
        """Switch to a new context: flush, reconnect, block for the new snapshot."""
        self.flush()
        if self._stream is not None:
            self._stream.close()
        with self._lock:
            self._store = {}
            self._context = context
        self._ready.clear()
        self._start_stream()
        return self.wait_for_initialization(timeout)

    def close(self) -> None:
        """Stop streaming and flush a final time."""
        self._closed.set()
        if self._stream is not None:
            self._stream.close()
        self.flush()

File: sdk/go/cached.go
package flagit

import (
	"context"
	"sync"
	"time"
)

// CachedOptions configures a CachedClient.
type CachedOptions struct {
	// BaseURL is the service root, e.g. "http://localhost:8080".
	BaseURL string
	// SDKKey is an environment's SDK key.
	SDKKey string
	// Context is the evaluation context this client streams and reads for.
	Context EvalContext
	// FlushInterval is how often buffered usage events are sent. 0 uses the
	// default (5s); a negative value disables the background flush.
	FlushInterval time.Duration
	// DisableEvents turns off usage reporting entirely.
	DisableEvents bool
	// OnError is called on a stream or flush error.
	OnError func(error)
}

// CachedClient mirrors the LaunchDarkly-style pattern: it streams flag values
// into an in-memory store once, then serves synchronous local reads with ZERO
// network per read, and reports usage as batched summary events. Use this for a
// long-lived app; use Client for one-off calls.
//
// All methods are safe for concurrent use.
type CachedClient struct {
	base       *Client
	trackEvent bool
	onError    func(error)

	mu        sync.RWMutex
	store     map[string]Evaluation
	ready     bool
	readyCh   chan struct{}
	current   EvalContext
	stream    *Stream
	listeners map[int]func()
	nextID    int
	counts    map[string]map[int]int

	cancel context.CancelFunc
	wg     sync.WaitGroup
}

// NewCachedClient creates a CachedClient and starts streaming immediately.
func NewCachedClient(opts CachedOptions) *CachedClient {
	onError := opts.OnError
	if onError == nil {
		onError = func(error) {}
	}
	ctx, cancel := context.WithCancel(context.Background())
	c := &CachedClient{
		base:       New(Options{BaseURL: opts.BaseURL, SDKKey: opts.SDKKey}),
		trackEvent: !opts.DisableEvents,
		onError:    onError,
		store:      map[string]Evaluation{},
		readyCh:    make(chan struct{}),
		current:    opts.Context,
		listeners:  map[int]func(){},
		counts:     map[string]map[int]int{},
		cancel:     cancel,
	}
	c.startStream(ctx)

	interval := opts.FlushInterval
	if interval == 0 {
		interval = 5 * time.Second
	}
	if interval > 0 && c.trackEvent {
		c.wg.Add(1)
		go func() {
			defer c.wg.Done()
			ticker := time.NewTicker(interval)
			defer ticker.Stop()
			for {
				select {
				case <-ctx.Done():
					return
				case <-ticker.C:
					c.Flush(context.Background())
				}
			}
		}()
	}
	return c
}

func (c *CachedClient) startStream(ctx context.Context) {
	c.stream = c.base.Stream(ctx, c.current, func(flags map[string]Evaluation) {
		c.mu.Lock()
		c.store = flags
		if !c.ready {
			c.ready = true
			close(c.readyCh)
		}
		listeners := make([]func(), 0, len(c.listeners))
		for _, l := range c.listeners {
			listeners = append(listeners, l)
		}
		c.mu.Unlock()
		for _, l := range listeners {
			l()
		}
	}, &StreamOptions{OnError: c.onError})
}

// WaitForInitialization blocks until the initial snapshot arrives or ctx is done.
func (c *CachedClient) WaitForInitialization(ctx context.Context) error {
	c.mu.RLock()
	ch := c.readyCh
	c.mu.RUnlock()
	select {
	case <-ch:
		return nil
	case <-ctx.Done():
		return ctx.Err()
	}
}

// Initialized reports whether the initial snapshot has loaded.
func (c *CachedClient) Initialized() bool {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.ready
}

// OnChange subscribes to store updates and returns an unsubscribe function.
func (c *CachedClient) OnChange(listener func()) func() {
	c.mu.Lock()
	id := c.nextID
	c.nextID++
	c.listeners[id] = listener
	c.mu.Unlock()
	return func() {
		c.mu.Lock()
		delete(c.listeners, id)
		c.mu.Unlock()
	}
}

func (c *CachedClient) read(flagKey string) (Evaluation, bool) {
	c.mu.RLock()
	ev, ok := c.store[flagKey]
	c.mu.RUnlock()
	if ok && c.trackEvent {
		c.mu.Lock()
		byVar := c.counts[flagKey]
		if byVar == nil {
			byVar = map[int]int{}
			c.counts[flagKey] = byVar
		}
		byVar[ev.Variation]++
		c.mu.Unlock()
	}
	return ev, ok
}

// BoolVariation returns the flag's boolean value from the local store, or
// fallback before init / on type mismatch. It performs no network I/O.
func (c *CachedClient) BoolVariation(flagKey string, fallback bool) bool {
	if ev, ok := c.read(flagKey); ok {
		if v, ok := ev.Value.(bool); ok {
			return v
		}
	}
	return fallback
}

// StringVariation returns the flag's string value from the local store, or fallback.
func (c *CachedClient) StringVariation(flagKey string, fallback string) string {
	if ev, ok := c.read(flagKey); ok {
		if v, ok := ev.Value.(string); ok {
			return v
		}
	}
	return fallback
}

// Float64Variation returns the flag's numeric value from the local store, or fallback.
func (c *CachedClient) Float64Variation(flagKey string, fallback float64) float64 {
	if ev, ok := c.read(flagKey); ok {
		if v, ok := ev.Value.(float64); ok {
			return v
		}
	}
	return fallback
}

// AllFlags returns a snapshot of all cached evaluations.
func (c *CachedClient) AllFlags() map[string]Evaluation {
	c.mu.RLock()
	defer c.mu.RUnlock()
	out := make(map[string]Evaluation, len(c.store))
	for k, v := range c.store {
		out[k] = v
	}
	return out
}

// Flush sends buffered usage events now.
func (c *CachedClient) Flush(ctx context.Context) error {
	c.mu.Lock()
	if len(c.counts) == 0 {
		c.mu.Unlock()
		return nil
	}
	summary := EventSummary{Flags: map[string]FlagCounters{}}
	for flagKey, byVar := range c.counts {
		counters := make([]Counter, 0, len(byVar))
		for variation, count := range byVar {
			counters = append(counters, Counter{Variation: variation, Count: count})
		}
		summary.Flags[flagKey] = FlagCounters{Counters: counters}
	}
	c.counts = map[string]map[int]int{}
	c.mu.Unlock()

	if err := c.base.SendEvents(ctx, summary); err != nil {
		c.onError(err)
		return err
	}
	return nil
}

// Identify switches to a new context: it flushes, reconnects the stream, and
// blocks until the new snapshot arrives or ctx is done.
func (c *CachedClient) Identify(ctx context.Context, ec EvalContext) error {
	c.Flush(ctx)
	c.stream.Close()

	c.mu.Lock()
	c.store = map[string]Evaluation{}
	c.ready = false
	c.readyCh = make(chan struct{})
	c.current = ec
	c.mu.Unlock()

	c.startStream(context.Background())
	return c.WaitForInitialization(ctx)
}

// Close stops streaming, flushes a final time, and releases resources.
func (c *CachedClient) Close() error {
	c.cancel()
	if c.stream != nil {
		c.stream.Close()
	}
	c.wg.Wait()
	return c.Flush(context.Background())
}
